File: Intro to Algorithms/PA2/Reference Scripts/Matrix.java
//Galen Turoci
//gturoci
//pa3

class Matrix {
	
	private class Entry {
		//Fields
		int column;
		double value;
		
		//Constructor
		Entry(int column, double value) {
			this.column = column;
			this.value = value;
		}
		
		// toString():  overrides Object's toString() method - NOT SURE HOW TO WRITE THIS ONE
		public String toString() {
			String Entrystring = "";
			Entrystring += ("(" + column + ", " + value + ")");
			return Entrystring;
		}
		
		//equals(Object x)
		//Tells us if this entry and the entered entry are the same
		//overrides Object's equals() method
		public boolean equals(Object x) {
			boolean eq = false;
			Entry that;
			if (x instanceof Entry) {
				that = (Entry) x;
				eq = (this.value ==that.value);
			}
			return eq;
		}
		
	}
	
	//Fields
	private Entry column;
	private Entry value;
	private int NNZ;
	private List[] Rows;
	
	//Constructor--------------------------------------------
	
	//Makes a new n x n zero Matrix
	//Pre: n >= 1
	Matrix(int n) {
		if (n < 1)
			throw new RuntimeException("Matrix Error: Cannot construct a Matrix with no entries!");
		Rows = new List[n];
		for (int i = 0; i < n; i++) 
			Rows[i] = new List();
		NNZ = 0;
	}
	
	//Access Functions----------------------------------------
	
	//getSize()
	//Returns the size of the array of lists that is this matrix
	int getSize() {
		return Rows.length;
	}
	
	//getNNZ()
	//Returns the number of non-0 entries in this matrix
	int getNNZ() {
		return NNZ;
	}
	
	//equals(Object x)
	//Tells us if this matrix and that matrix are twinsies
	public boolean equals(Object x) {
		boolean eq = false;
		Matrix that;
		Entry e, ex;
		int y, z, i;
		double a, b;
		if (x instanceof Matrix) {
			that = (Matrix) x;
			
			eq = (this.getSize() == that.getSize());
			if (!eq)
				return eq;
			
			eq = (this.getNNZ() == that.getNNZ());
			if (!eq)
				return eq;
			
			for (i = 0; i < Rows.length; i++) {
				List ithRow = Rows[i];
				List ithRowx = that.Rows[i];
				
				eq = (ithRow.equals(ithRowx));
				if (!eq)
					return eq;
				
				if (ithRow.length() == 0 && ithRowx.length() == 0)
					continue;
			}
		}
		return eq;
	}
	
	//Manipulation procedures------------------
	
	//makeZero()
	//Sets this matrix to the 0 state
	void makeZero() {
		List ithRow = new List();
		for (int i = 0; i < Rows.length; i++) {
			ithRow = Rows[i];
			if (ithRow.length() == 0)
				continue;
			
			ithRow.clear();
		}
		NNZ = 0;
	}
	
	//copy()
	//Clones this matrix
	Matrix copy() {
		Matrix newM = new Matrix(Rows.length);
		List ithRow = new List();
		Entry e = null;
		int c, i;
		double x;
		
		for (i = 0; i < Rows.length; i++) {	
			ithRow = Rows[i];
			if (ithRow.length() == 0)
				continue;
			
			ithRow.moveFront();
			while (ithRow.index() >= 0) {
				e = (Entry) ithRow.get();
				c = e.column;
				x = e.value;
				newM.changeEntry(i+1, c, x);
				ithRow.moveNext();
			}
		}
		return newM;
	}
	
	//changeEntry(int i, int j, double x)
	//changes ith row, jth column of this matrix to x
	//pre: 1 <= i <= getSize(), 1 <= j <= getSize()
	void changeEntry(int i, int j, double x) {
		if (1 > i || i > this.getSize() || 1 > j || j > this.getSize())
			throw new RuntimeException("Matrix Error: Can't change what isn't there!");
		
		List ithRow = Rows[i-1];
		int c;
		Entry e;
		
		if (ithRow.length() == 0 && x != 0.0) {
			ithRow.append(new Entry(j, x));
			NNZ++;
		} else if (ithRow.length() != 0) {
			ithRow.moveFront();
			e = (Entry) ithRow.get();
			c = e.column;
			while (c < j && ithRow.index() != ithRow.length() - 1) {
				ithRow.moveNext();
				e = (Entry) ithRow.get();
				c = e.column;
			}
			if (c < j && x != 0.0) {
				ithRow.insertAfter(new Entry(j, x));
				NNZ++;
			} else if (c == j) {
				if (x != 0.0)
					e.value = x;
				else {
					ithRow.delete();
					NNZ--;
				}
			} else if (x != 0.0) {
				ithRow.insertBefore(new Entry(j, x));
				NNZ++;
			}
			ithRow.moveBack();
			ithRow.moveNext();
		}
	}
	
	//scalarMult(double x)
	//returns a new Matrix that is the scalar product of this Matrix with x
	Matrix scalarMult(double x) {
		Matrix newM = new Matrix(Rows.length);
		List ithRow = new List();
		Entry e;
		int c, i;
		double y;
		
		for (i = 0; i < Rows.length; i++) {
			ithRow = Rows[i];
			if (ithRow.length() == 0)
				continue;
			
			ithRow.moveFront();
			while (ithRow.index() >= 0) {
				e = (Entry) ithRow.get();
				c = e.column;
				y = e.value;
				newM.changeEntry(i+1, c, (x*y));
				ithRow.moveNext();
			}
		}
		return newM;
	}
	
	//add(Matrix M)
	//adds this Matrix with Matrix M
	//pre: getSize() == M.getSize()
	Matrix add(Matrix M) {
		if (this.getSize() != M.getSize())
			throw new RuntimeException("Matrix Error: Can't add Matrices of different sizes!");
		
		if (this.equals(M))
			return M.scalarMult(2.0);
		
		Matrix newM = new Matrix(Rows.length);
		List ithRow = new List();
		List ithRowM = new List();
		Entry e;
		Entry eM;
		int c, cM, i;
		double x = 0.0, y = 0.0;
		
		for (i = 0; i < Rows.length; i++) {
			ithRow = Rows[i];
			ithRowM = M.Rows[i];
			if (ithRow.length() == 0 && ithRowM.length() == 0)
				continue;
			
			ithRow.moveFront();
			ithRowM.moveFront();
			while (ithRow.index() >= 0 || ithRowM.index() >= 0) {
				c = 0;
				cM = 0;
				if (ithRow.length() != 0 && ithRow.index() != -1) {
					e = (Entry) ithRow.get();
					c = e.column;
					x = e.value;
				}
				if (ithRowM.length() != 0 && ithRowM.index() != -1) {
					eM = (Entry) ithRowM.get();
					cM = eM.column;
					y = eM.value;
				}
				
				if (c == cM) {
					newM.changeEntry(i+1, c, (x+y));
					ithRow.moveNext();
					ithRowM.moveNext();
				} else if (c != 0 && (c < cM || cM == 0)) {
					newM.changeEntry(i+1, c, x);
					ithRow.moveNext();
				} else {
					newM.changeEntry(i+1, cM, y);
					ithRowM.moveNext();
				}
			}
		}
		return newM;
	}
	
	//sub(Matrix M)
	//subtracts this matrix with Matrix M
	//pre: getSize() == M.getSize()
	Matrix sub(Matrix M) {
		if (this.getSize() != M.getSize())
			throw new RuntimeException("Matrix Error: Can't sub Matrices of different sizes!");
		
		Matrix newM = new Matrix(Rows.length);
		if (this.equals(M))
			return newM;
		
		List ithRow = new List();
		List ithRowM = new List();
		Entry e;
		Entry eM;
		int c, cM, i;
		double x = 0.0, y = 0.0;
		
		for (i = 0; i < Rows.length; i++) {
			ithRow = Rows[i];
			ithRowM = M.Rows[i];
			if (ithRow.length() == 0 && ithRowM.length() == 0)
				continue;
			
			ithRow.moveFront();
			ithRowM.moveFront();
			while (ithRow.index() >= 0 || ithRowM.index() >= 0) {
				c = 0;
				cM = 0;
				if (ithRow.length() != 0 && ithRow.index() != -1) {
					e = (Entry) ithRow.get();
					c = e.column;
					x = e.value;
				}
				if (ithRowM.length() != 0 && ithRowM.index() != -1) {
					eM = (Entry) ithRowM.get();
					cM = eM.column;
					y = eM.value;
				}
				
				if (c == cM) {
					newM.changeEntry(i+1, c, (x-y));
					ithRow.moveNext();
					ithRowM.moveNext();
				} else if (c != 0 && (c < cM || cM == 0)) {
					newM.changeEntry(i+1, c, -x);
					ithRow.moveNext();
				} else {
					newM.changeEntry(i+1, cM, -y);
					ithRowM.moveNext();
				}
			}
		}
		return newM;
	}
	
	//transpose()
	//Returns a new matrix that is the transpose of this one
	Matrix transpose() {
		Matrix newM = new Matrix(Rows.length);
		List ithRow = new List();
		Entry e;
		int x, i;
		double y;
		
		for (i = 0; i < Rows.length; i++) {
			ithRow = Rows[i];
			if (ithRow.length() == 0)
				continue;
			
			ithRow.moveFront();
			while (ithRow.index() >= 0) {
				e = (Entry) ithRow.get();
				x = e.column;
				y = e.value;
			
				newM.changeEntry(x, i+1, y);
				
				ithRow.moveNext();
			}
		}
		return newM;
	}
	
	//mult(Matrix M)
	//Returns a new matrix that is the product of this matrix with M
	//pre: getSize() == M.getSize()
	Matrix mult(Matrix M) {
		if (this.getSize() != M.getSize())
			throw new RuntimeException("Matrix Error: Can't mult Matrices of different sizes!");
		
		Matrix newM = new Matrix(Rows.length);
		List ithRow = new List();
		List ithRowM = new List();
		int a = Rows.length, b, c, i, j;
		double newY;
		Entry[] e = new Entry[a];
		Entry[] eM = new Entry[a];
		double[] y = new double[a];
		double[] yM = new double[a];
		
		for (i = 0; i < a; i++) {
			ithRow = Rows[i];
			if (ithRow.length() == 0)
				continue;
			
			for (b = 0; b < y.length; b++)
				y[b] = 0.0;
			
			ithRow.moveFront();
			for (b = 0; b < ithRow.length(); b++) {
				e[b] = (Entry) ithRow.get();
				y[e[b].column-1] = e[b].value;
				ithRow.moveNext();
			}
			
			for (j = 1; j <= a; j++) {
				for (b = 0; b < a; b++) {
					ithRowM = M.Rows[b];
					if (ithRowM.length() == 0)
						continue;
				
					ithRowM.moveFront();
					while (ithRowM.index() != j-1 && ithRowM.index() != ithRowM.length() - 1)
						ithRowM.moveNext();
				
					eM[b] = (Entry) ithRowM.get();
					yM[b] = eM[b].value;
				
					if (eM[b].column != j)
						yM[b] = 0.0;
				}
			
				c = 0;
				newY = 0.0;
				while (c < a) {
					newY += (y[c] * yM[c]);
					c++;
				}
			
				newM.changeEntry(i+1, j, newY);
			}
		}
		return newM;
	}
	
	//Other functions-------------------------------------------
	
	//toString()
	//overrides Object's toString() method
	public String toString() {
		StringBuffer sb = new StringBuffer();
		for (int i = 0; i < Rows.length; i++) {
			if (Rows[i].length() != 0) {
				sb.append((i+1) + ": ");
				sb.append(Rows[i]);
				sb.append("\n");
			}
		}
		return new String(sb);	
	}
	
}

File: Intro to Algorithms/PA2/Scripts/Matrix.c
//Galen Turoci
//gturoci
//PA2

#include<stdio.h>
#include<stdlib.h>
#include "Matrix.h"
#include "List.h"


//Structs ---------------------------------------------------------------------


//Private NodeObj type
typedef struct EntryObj {
	int column;
	double value;
} EntryObj;

//Private Entry type
typedef EntryObj* Entry;

//Private MatrixObj type
typedef struct MatrixObj {
	int NNZ;
	int size;
	List* Rows;
} MatrixObj;


//Constructors-Destructors ----------------------------------------------------


//newEntry()
//Returns a reference to a new Entry object
Entry newEntry(int x, double y) {
	Entry E = malloc(sizeof(EntryObj));
	E->column = x;
	E->value = y;
	
	return(E);
}

//freeEntry()
//Frees all heap memory associated with Entry *pE, and sets *pE to NULL.
void freeEntry(Entry* pE) {
	if (pE != NULL && *pE != NULL) {
		free(*pE);
		*pE = NULL;
	}
}

//newMatrix()
//Returns reference to new empty Matrix object in the zero state.
Matrix newMatrix(int n) {
	Matrix M = malloc(sizeof(MatrixObj));
	
	//Initialize Rows
	M->Rows = calloc(n, sizeof(List));
	for (int i = 0; i < n; i++)
		M->Rows[i] = newList();
	
	//Initialize NNZ
	M->NNZ = 0;
	
	//We know our size now
	M->size = n;
	
	return(M);
}

//freeMatrix()
//Frees all heap memory associated with Matrix *pM, the Lists therein,
//and sets *pM to NULL.
//This calls makeZero, which calls clear, which SHOULD be freeing the memory of
//my Lists (LOL), then frees the Matrix itself and sets its pointer to NULL
void freeMatrix(Matrix* pM) {
	if (pM != NULL && *pM != NULL) { 
		if (size(*pM) > 0)
			makeZero(*pM);
		free(*pM);
		*pM = NULL;
	}
}


// Access functions -----------------------------------------------------------

//size()
//Returns the size of square Matrix M.
int size(Matrix M) {
	if (M == NULL) {
		printf("Matrix Error: calling size() on NULL Matrix reference\n");
		exit(1);
	}
	
	return (M->size);
}

//NNZ()
//Returns the number of non-zero elements in M.
int NNZ(Matrix M) {
	if (M == NULL) {
		printf("Matrix Error: calling NNZ() on NULL Matrix reference\n");
		exit(1);
	}
	
	return M->NNZ;
}

//equals()
//Returns true (1) if matrices A and B are equal, false (0) otherwise.
int equals(Matrix A, Matrix B) {
	if (A == NULL || B == NULL) {
		printf("Matrix Error: calling equals() on NULL Matrix reference\n");
		exit(1);
	}
	
	Matrix copyB = newMatrix(size(B));
	copyB = copy(B);
	
	Entry eA, eB;
	int y, z, i, eq = 0;
	double a, b;
	
	eq = ((size(A) == size(copyB)) && (NNZ(A) == NNZ(copyB)));
	if (eq == 0)
		return eq;
	
	for (i = 0; i < size(A); i++) {
		List ithRowA = A->Rows[i];
		List ithRowB = copyB->Rows[i];
		
		if (length(ithRowA) == 0 && length(ithRowB) == 0)
			continue;
		
		moveFront(ithRowA);
		moveFront(ithRowB);
		while (eq && (index(ithRowA) >= 0) && (index(ithRowB) >= 0)) {
			eA = (Entry) get(ithRowA);
			eB = (Entry) get(ithRowB);
			y = eA->column;
			z = eB->column;
			a = eA->value;
			b = eB->value;
			
			eq = ((y == z) && (a == b));
			
			moveNext(ithRowA);
			moveNext(ithRowB);
		}
	}

	return eq;
}


//Manipulation procedures ----------------------------------------------------

//makeZero()
//Resets M to the zero state
void makeZero(Matrix M) {
	if (M == NULL) {
		printf("Matrix Error: calling makeZero() on NULL Matrix reference\n");
		exit(1);
	}
	
	List ithRow = newList();
	for (int i = 0; i < size(M); i++) {
		ithRow = M->Rows[i];
		if (length(ithRow) == 0)
			continue;
		
		clear(ithRow);
	}
	M->NNZ = 0;
}

//changeEntry()
//Changes the ithg row, jth column of M to the value x.
//Pre: 1 <= i <= size(M), 1 <= j <= size(M)
void changeEntry(Matrix M, int i, int j, double x) {
	if (M == NULL) {
		printf("Matrix Error: calling changeEntry() on NULL Matrix reference\n");
		exit(1);
	} if (1 > i || i > size(M) || 1 > j || j > size(M)) {
		printf("Matrix Error: Can't change what isn't there!\n");
		exit(1);
	}
	
	List ithRow = M->Rows[i-1];
	int c;
	Entry e;
	
	if (length(ithRow) == 0 && x != 0.0) {
		append(ithRow, newEntry(j, x));
		M->NNZ++;
	} else if (length(ithRow) != 0) {
		moveFront(ithRow);
		e = (Entry) get(ithRow);
		c = e->column;
		while (c < j && index(ithRow) != length(ithRow) - 1) {
			moveNext(ithRow);
			e = (Entry) get(ithRow);
			c = e->column;
		}
		if (c < j && x != 0.0) {
			insertAfter(ithRow, newEntry(j, x));
			M->NNZ++;
		} else if (c == j) {
			if (x != 0.0)
				e->value = x;
			else {
				delete(ithRow);
				M->NNZ--;
			}
		} else if (x != 0.0) {
			insertBefore(ithRow, newEntry(j, x));
			M->NNZ++;
		}
		moveBack(ithRow);
		moveNext(ithRow);
	}
}


//Matrix Arithmetic operations ------------------------------------------------


//copy()
//Returns a reference to a new Matrix object having the same entries as A.
Matrix copy(Matrix A) {
	if (A == NULL) {
		printf("Matrix Error: calling copy() on NULL Matrix reference\n");
		exit(1);
	}
	
	Matrix newM = newMatrix(size(A));
	List ithRow = newList();
	Entry e;
	int c, i;
	double x;
	
	for (i = 0; i < size(A); i++) {	
		ithRow = A->Rows[i];
		if (length(ithRow) == 0)
			continue;
		
		moveFront(ithRow);
		while (index(ithRow) >= 0) {
			e = (Entry) get(ithRow);
			c = e->column;
			x = e->value;
			changeEntry(newM, i+1, c, x);
			moveNext(ithRow);
		}
	}
	return newM;
}

//transpose()
// Returns a reference to a new Matrix object representing the transpose of A.
Matrix transpose(Matrix A) {
	if (A == NULL) {
		printf("Matrix Error: calling transpose() on NULL Matrix reference\n");
		exit(1);
	}
	
	Matrix newM = newMatrix(size(A));
	List ithRow = newList();
	Entry e;
	int x, i;
	double y;
	
	for (i = 0; i < size(A); i++) {
		ithRow = A->Rows[i];
		if (length(ithRow) == 0)
			continue;
		
		moveFront(ithRow);
		while (index(ithRow) >= 0) {
			e = (Entry) get(ithRow);
			x = e->column;
			y = e->value;
		
			changeEntry(newM, x, i+1, y);
			
			moveNext(ithRow);
		}
	}
	return newM;
}

//scalarMult()
//Returns a reference to a new Matrix object representing xA.
Matrix scalarMult(double x, Matrix A) {
	if (A == NULL) {
		printf("Matrix Error: calling scalarMult() on NULL Matrix reference\n");
		exit(1);
	}
	
	Matrix newM = newMatrix(size(A));
	List ithRow = newList();
	Entry e;
	int c, i;
	double y;
	
	for (i = 0; i < size(A); i++) {
		ithRow = A->Rows[i];
		if (length(ithRow) == 0)
			continue;
		
		moveFront(ithRow);
		while (index(ithRow) >= 0) {
			e = (Entry) get(ithRow);
			c = e->column;
			y = e->value;
			changeEntry(newM, i+1, c, (x*y));
			moveNext(ithRow);
		}
	}
	return newM;
}

//sum()
//Returns a reference to a new Matrix object representing A+B.
//pre: size(A)==size(B)
Matrix sum(Matrix A, Matrix B) {
	if (A == NULL || B == NULL) {
		printf("Matrix Error: calling sum() on NULL Matrix reference\n");
		exit(1);
	} if (size(A) != size(B)) {
		printf("Matrix Error: Can't add Matrices of different sizes!");
		exit(1);
	}
	
	if (equals(A, B))
		return scalarMult(2.0, A);
	
	Matrix newM = newMatrix(size(A));
	List ithRowA = newList();
	List ithRowB = newList();
	Entry eA;
	Entry eB;
	int cA, cB, i;
	double xA = 0.0, xB = 0.0;
	
	for (i = 0; i < size(A); i++) {
		ithRowA = A->Rows[i];
		ithRowB = B->Rows[i];
		if (length(ithRowA) == 0 && length(ithRowB) == 0)
			continue;
		
		moveFront(ithRowA);
		moveFront(ithRowB);
		while (index(ithRowA) >= 0 || index(ithRowB) >= 0) {
			cA = 0;
			cB = 0;
			if (length(ithRowA) != 0 && index(ithRowA) != -1) {
				eA = (Entry) get(ithRowA);
				cA = eA->column;
				xA = eA->value;
			}
			if (length(ithRowB) != 0 && index(ithRowB) != -1) {
				eB = (Entry) get(ithRowB);
				cB = eB->column;
				xB = eB->value;
			}
			
			if (cA == cB) {
				changeEntry(newM, i+1, cA, (xA+xB));
				moveNext(ithRowA);
				moveNext(ithRowB);
			} else if (cA != 0 && (cA < cB || cB == 0)) {
				changeEntry(newM, i+1, cA, xA);
				moveNext(ithRowA);
			} else {
				changeEntry(newM, i+1, cB, xB);
				moveNext(ithRowB);
			}
		}
	}
	return newM;
}

//diff()
//Returns a reference to a new Matrix object representing A-B.
//pre: size(A)==size(B)
Matrix diff(Matrix A, Matrix B) {
	if (A == NULL || B == NULL) {
		printf("Matrix Error: calling diff() on NULL Matrix reference\n");
		exit(1);
	} if (size(A) != size(B)) {
		printf("Matrix Error: Can't sub Matrices of different sizes!");
		exit(1);
	}
	
	Matrix newM = newMatrix(size(A));
	if (equals(A, B) == 1)
		return newM;
	
	List ithRowA = newList();
	List ithRowB = newList();
	Entry eA;
	Entry eB;
	int cA, cB, i;
	double xA = 0.0, xB = 0.0;
	
	for (i = 0; i < size(A); i++) {
		ithRowA = A->Rows[i];
		ithRowB = B->Rows[i];
		if (length(ithRowA) == 0 && length(ithRowB) == 0)
			continue;
		
		moveFront(ithRowA);
		moveFront(ithRowB);
		while (index(ithRowA) >= 0 || index(ithRowB) >= 0) {
			cA = 0;
			cB = 0;
			if (length(ithRowA) != 0 && index(ithRowA) != -1) {
				eA = (Entry) get(ithRowA);
				cA = eA->column;
				xA = eA->value;
			}
			if (length(ithRowB) != 0 && index(ithRowB) != -1) {
				eB = (Entry) get(ithRowB);
				cB = eB->column;
				xB = eB->value;
			}
		
			if (cA == cB) {
				changeEntry(newM, i+1, cA, (xA-xB));
				moveNext(ithRowA);
				moveNext(ithRowB);
			} else if (cA != 0 && (cA < cB || cB == 0)) {
				changeEntry(newM, i+1, cA, xA);
				moveNext(ithRowA);
			} else {
				changeEntry(newM, i+1, cB, -xB);
				moveNext(ithRowB);
			}
		}
	}
	return newM;
}

//product()
//Returns a reference to a new Matrix object representing AB
//pre: size(A)==size(B)
Matrix product(Matrix A, Matrix B) {
	if (A == NULL || B == NULL) {
		printf("Matrix Error: calling product() on NULL Matrix reference\n");
		exit(1);
	} if (size(A) != size(B)) {
		printf("Matrix Error: Can't mult Matrices of different sizes!");
		exit(1);
	}
	
	Matrix newM = newMatrix(size(A));
	List ithRowA = newList();
	List ithRowB = newList();
	int a = size(A), b, c, i, j, yAlength;
	double newY;
	Entry eA[a];
	Entry eB[a];
	double yA[a];
	double yB[a];
	
	yAlength = sizeof(yA)/sizeof(double);
	
	for (i = 0; i < a; i++) {
		ithRowA = A->Rows[i];
		if (length(ithRowA) == 0)
			continue;
		
		for (b = 0; b < yAlength; b++)
			yA[b] = 0.0;
		
		moveFront(ithRowA);
		for (b = 0; b < length(ithRowA); b++) {
			eA[b] = (Entry) get(ithRowA);
			yA[eA[b]->column-1] = eA[b]->value;
			moveNext(ithRowA);
		}
		
		for (j = 1; j <= a; j++) {
			for (b = 0; b < a; b++) {
				ithRowB = B->Rows[b];
				if (length(ithRowB) == 0)
					continue;
			
				moveFront(ithRowB);
				while (index(ithRowB) != j-1 && index(ithRowB) != length(ithRowB) - 1)
					moveNext(ithRowB);
			
				eB[b] = (Entry) get(ithRowB);
				yB[b] = eB[b]->value;
			
				if (eB[b]->column != j)
					yB[b] = 0.0;
			}
			
			c = 0;
			newY = 0.0;
			while (c < a) {
				newY += (yA[c] * yB[c]);
				c++;
			}
		
			changeEntry(newM, i+1, j, newY);
		}
	}
	return newM;	
}

//printMatrix()
//Prints a string representation of Matrix M to filestream out. Zero rows are
//not printed. Each non-zero is represented as one line consisting of the row
//number, followed by a colon, a space, then a space separated list of pairs
//"(col, val)" giving the column numbers and non-zero values in that row. The
//double val will be rounded to 1 decimal point.
void printMatrix(FILE* out, Matrix M) {
	if (M == NULL) {
		printf("Matrix Error: calling printMatrix() on NULL Matrix reference\n");
		exit(1);
	}
	
	List ithRow;
	Entry e;
	int c, i;
	double x;
	
	for (i = 0; i < size(M); i++) {
		ithRow = M->Rows[i];
		if (length(ithRow) != 0) {
			moveFront(ithRow);
			fprintf(out, "%d:", i+1);
			while (index(ithRow) >= 0) {
				e = (Entry) get(ithRow);
				c = e->column;
				x = e->value;
			
				fprintf(out, " (%d, %.1f)", c, x);
				moveNext(ithRow);
			}
			fprintf(out, "\n");
		}
	}
}